### PR TITLE
Add livepush to balena push

### PR DESCRIPTION
**Pull request by CameronDiver**
Depends-on: https://github.com/balena-io/balena-supervisor/pull/828

---
##### Contributor checklist

- [ ] Introduces security considerations
- [ ] Affects the development, build or deployment processes of the component

### CODE REVIEW

File: lib/actions/push.ts [hunk 1]
```diff
@@ -166,6 +176,14 @@ export const push: CommandDefinition<
 			description: stripIndent`
 				Path to a local YAML or JSON file containing Docker registry passwords used to pull base images`,
 		},
+		{
+			signature: 'live',
+			alias: 'l',
+			boolean: true,
+			description: stripIndent`
+				Start a live session after the push, which will wait for code changes, and synchronise them with
+				running containers. This mode is only valid when pushing to a local device.`,
```
Comment (pdcastro) [inline lib/actions/push.ts]: I suggest a short description here, with a reference, so it formats more nicely on the screen:
```
Start a live session after the push. See live push description above.
```

File: lib/utils/device/deploy.ts [hunk 2]
```diff
@@ -82,9 +85,18 @@ export async function deployToDevice(opts: DeviceDeployOptions): Promise<void> {
...
 	try {
 		const version = await api.getVersion();
+		globalLogger.logDebug(`Checking device version: ${version}`);
 		if (!semver.satisfies(version, '>=7.21.4')) {
 			exitWithExpectedError(versionError);
 		}
+		// FIXME: DO NOT MERGE until this version number has been updated
+		// with the version which the following PR ends up in the supervisor
+		// https://github.com/balena-io/balena-supervisor/pull/828
+		if (opts.live && !semver.satisfies(version, '>=1.0.0')) {
```
[the remark this answers is not in the document]
Comment (CameronDiver) [reply lib/utils/device/deploy.ts]: It should be, but I'd like to do more testing before locking this in, in case the supervisor needs any more patches.

**Comment by pdcastro** [comment]
Just a note, before I forget: the Travis builds are not labeled 'Required' but they are actually (still) required. They are (still) the ones that create the MacOS and Linux standalone release zips files in the [balena-cli repo's releases page](https://github.com/balena-io/balena-cli/releases). I think the reason why they are currently failing for this PR is that `npm run build` was not run on your laptop to generate the markdown docs before git-pushing. (Yes, maybe it would make sense for the CI to generate that even if `npm run build` wasn't run on the laptop...)

File: lib/actions/push.ts [hunk 3]
```diff
@@ -107,15 +107,25 @@ export const push: CommandDefinition<
...
+		or a local mode balena device.
+
+		When building on the balena cloud the given source directory will be sent to the
+		balena builder, and the build will proceed. This can be used as a drop-in
+		replacement for git push to deploy.
+
+		When building on a local mode device, the given source directory will be built
+		on the device, and the resulting containers will be run on the device. Logs will
+		be streamed back from the device as part of the same invocation. This requires
+		a device with a supervisor version of at least v7.21.0.
```
[the remark this answers is not in the document]
Comment (hedss) [reply lib/actions/push.ts]: I'd personally move this description to the `live` signature description below, and move a small two line sentence here that says 'A live session can be started after the push by passing the `--live` flag', or similar. Else we're starting a precedence for any major sub-feature.

File: lib/utils/device/live.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	buildContext: string;
+	composition: Composition;
+	buildTasks: BuildTask[];
+	docker: Dockerode;
+	api: DeviceAPI;
+	logger: Logger;
+	buildLogs: BuildLogs;
+}
+
+export class LivepushManager {
```
[the remark this answers is not in the document]
Comment (hedss) [reply lib/utils/device/live.ts]: I think the reason we haven't is that we've tried to port a lot over to TS relatively quickly. The entire codebase could probably do with going over and auditing. But, time...

File: lib/utils/device/live.ts [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+			`Device state not settled, retrying in ${DEVICE_STATUS_SETTLE_CHECK_INTERVAL}ms`,
+		);
+		await Bluebird.delay(DEVICE_STATUS_SETTLE_CHECK_INTERVAL);
+		await this.awaitDeviceStateSettle();
+	}
+
+	private async handleFSEvent(fsEvent: ContextEvent): Promise<void> {
+		// TODO: If there's a dockerfile event, we must perform a rebuild
+		this.logger.logDebug(
+			`Got an filesystem event for service: ${
```
Comment (hedss) [inline lib/utils/device/live.ts]: `Got a filesystem`

File: lib/utils/device/live.ts [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+				fsEvent.serviceName
+			}. Event: ${JSON.stringify(fsEvent)}`,
+		);
+
+		let updates: string[] = [];
+		let deletes: string[] = [];
+		switch (fsEvent.type) {
+			case 'add':
```
Comment (hedss) [inline lib/utils/device/live.ts]:
```suggestion
			case 'add':
			case 'change':
				updates = [fsEvent.filename];
				break;
```

File: typings/global.d.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+interface Dictionary<T> {
```
Comment (hedss) [inline typings/global.d.ts]: I find this amusing. :)

File: lib/utils/device/live.ts [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+		process.removeAllListeners('SIGINT');
+		process.on('SIGINT', async () => {
+			this.logger.logLivepush('Cleaning up device...');
+			await Promise.all(
+				_.map(this.containers, container => {
+					container.livepush.cleanupIntermediateContainers();
+				}),
+			);
+
+			process.exit(0);
```
Comment (hedss) [inline lib/utils/device/live.ts]: `process.exitCode = 0` instead maybe?

Comment (CameronDiver) [reply lib/utils/device/live.ts]: We need to actually force an exit, as we caught the interrupt.

Comment (hedss) [reply lib/utils/device/live.ts]: Ah, good point.

**Pull request merged**